### PR TITLE
Upgrade/layers

### PR DESCRIPTION
**Pull request by khein-pdftron**
Added nested layers.

### CODE REVIEW

File: src/components/LayersPanel/LayersPanel.js [hunk 1]
```diff
@@ -39,4 +49,9 @@ const mapStateToProps = state => ({
   isDisabled: selectors.isElementDisabled(state, 'layersPanel'),
 });
 
-export default connect(mapStateToProps)(withTranslation()(LayersPanel));
+export default connect(
+  mapStateToProps,
+  {
+    setLayers: actions.setLayers,
```
Comment (ZhijieZhang) [inline src/components/LayersPanel/LayersPanel.js]: Could you please add this to the propTypes?

File: src/components/Layer/Layer.js [hunk 2]
```diff
@@ -1,38 +1,101 @@
...
+      newChildLayer.visible = false;
+      newChildLayer = this.unCheckChildren(newChildLayer);
+      newLayer.children[i] = newChildLayer;
+    });
+    return newLayer;
+  }
 
   onChange = e => {
-    const { index, layers } = this.props;
+    const { updateLayer, layer } = this.props;
```
Comment (ZhijieZhang) [inline src/components/Layer/Layer.js]: Can you add `updateLayer` to propTypes as well. Thanks!

File: src/components/Layer/Layer.js [hunk 3]
```diff
@@ -1,38 +1,101 @@
...
+            onChange={this.onChange}
+            checked={layer.visible}
+          />
+        </div>
+        {hasSubLayers && isExpanded && (
+          <div className="sub-layers">
+            {layer.children.map((subLayer, i) => (
+              <Layer
+                key={i}
+                index={i}
```
Comment (ZhijieZhang) [inline src/components/Layer/Layer.js]: Do we still need to pass `index` and `layers` down as props?

File: src/components/Layer/Layer.js [hunk 4]
```diff
@@ -1,38 +1,101 @@
 import React from 'react';
 import PropTypes from 'prop-types';
-import { connect } from 'react-redux';
-
-import core from 'core';
 
+import Icon from 'components/Icon';
 import Input from 'components/Input';
 
+import core from 'core';
```
Comment (ZhijieZhang) [inline src/components/Layer/Layer.js]: Can you remove this as it's not used in this file

File: src/redux/actions/internalActions.js [hunk 5]
```diff
@@ -233,10 +233,16 @@ export const setOutlines = outlines => ({
...
-});
+export const setLayers = layers => (dispatch) => {
+  dispatch({
+    type: 'SET_LAYERS',
+    payload: { layers },
+  });
+  const doc = core.getDocument();
+  doc.setLayersArray(layers);
+  window.docViewer.refreshAll();
+  window.docViewer.updateView();
```
Comment (ZhijieZhang) [inline src/redux/actions/internalActions.js]: Do you think we should put this logic in a `componentDidUpdate` in `LayersPanel.js`? I guess that doesn't really make a difference...

File: src/components/LayersPanel/LayersPanel.js (hunk 1 above)
Comment (khein-pdftron) [reply src/components/LayersPanel/LayersPanel.js]: Done

File: src/components/Layer/Layer.js (hunk 4 above)
Comment (khein-pdftron) [reply src/components/Layer/Layer.js]: Done

File: src/redux/actions/internalActions.js (hunk 5 above)
Comment (khein-pdftron) [reply src/redux/actions/internalActions.js]: Done

File: src/components/Layer/Layer.js (hunk 3 above)
Comment (khein-pdftron) [reply src/components/Layer/Layer.js]: Nope. Removed.

File: src/components/Layer/Layer.js (hunk 2 above)
Comment (khein-pdftron) [reply src/components/Layer/Layer.js]: Done

File: src/components/LayersPanel/LayersPanel.js (hunk 1 above)
Comment (khein-pdftron) [reply src/components/LayersPanel/LayersPanel.js]: Done

**Pull request merged**